adafruit_tinylora/adafruit_tinylora.py

```diff
@@ -0,0 +1,405 @@
+        self._rfm_mid = None
+        self._rfm_lsb = None
+        # Set datarate registers
+        self._sf = None
+        self._bw = None
+        self._modemcfg = None
+        self.set_datarate("SF7BW125")
+        # Set regional frequency plan
+        if 'US' in ttn_config.country:
+            from adafruit_tinylora.ttn_usa import TTN_FREQS
+            self._frequencies = TTN_FREQS
+        elif adafruit_tinylora.ttn_config.country == 'AS':
+            from adafruit_tinylora.ttn_as import TTN_FREQS
+            self._frequencies = TTN_FREQS
+        elif adafruit_tinylora.ttn_config.country == 'AU':
+            from adafruit_tinylora.ttn_au import TTN_FREQS
+            self._frequencies = TTN_FREQS
+        elif adafruit_tinylora.ttn_config.country == 'EU':
+            from adafruit_tinylora.ttn_eu import TTN_FREQS
+            self._frequencies = TTN_FREQS
+        else:
+            print("Country Code Incorrect/Unsupported")
+        # Set Channel Number
+        self._channel = channel
+        self._tx_random = randint(0, 7)
+        if self._channel is not None:
+            # set single channel
+            self.set_channel(self._channel)
+        # Init FrameCounter
+        self.frame_counter = 0
+        # Verify the version of the RFM module
+        self._version = self._read_u8(0x42)
+        if self._version != 18:
+            print("Error Detecting RFM95W.")
+            print(" Check your wiring.")
+        # Set RFM to Sleep Mode
+        self._write_u8(0x01, _MODE_SLEEP)
+        # Set RFM to LoRa mode
+        self._write_u8(0x01, _MODE_LORA)
+        # Set Max. Power
+        self._write_u8(0x09, 0xFF)
+        # Set RX Timeout
+        self._write_u8(0x1F, 0x25)
+        # Preamble Length = 8
+        # Setup preamble (0x0008 + 4)
+        self._write_u8(_REG_PREAMBLE_MSB, 0x00)
+        self._write_u8(_REG_PREAMBLE_LSB, 0x08)
+        # Low datarate optimization off AGC auto on
+        self._write_u8(0x26, 0x0C)
+        # Set LoRa sync word
+        self._write_u8(0x39, 0x34)
+        # Set IQ to normal values
+        self._write_u8(0x33, 0x27)
+        self._write_u8(0x3B, 0x1D)
+        # Set FIFO pointers
+        # TX base adress
+        self._write_u8(0x0E, 0x80)
+        # Rx base adress
+        self._write_u8(0x0F, 0x00)
+        # Give the lora object ttn configuration
+        self._ttn_config = ttn_config
+
+    def send_data(self, data, data_length, frame_counter, timeout=2):
+        """Function to assemble and send data
+           :param data: data to send
+           :param data_length: length of data to send
+           :param frame_counter: frame counter variable, declared in code.py.
+           :param timeout: TxDone wait time, default is 2.
+        """
+        # data packet
+        enc_data = bytearray(data_length)
+        lora_pkt = bytearray(64)
+        # copy bytearray into bytearray for encryption
+        for i in range(0, data_length):
+            enc_data[i] = data[i]
```

Comment:
This copy seems unnecessary. You can create is as the return from the call to `aes.encrypt`. (see below)

Reply:
Copy to enc_data is required for sending (non-changing) data, running it through aes.encrypt will destroy the data, so a copy is required.